src/plugins/procdump/procdump.cpp

```diff
@@ -0,0 +1,832 @@
+ * and redistribute Covered Software under those terms instead of these.   *
+ *                                                                         *
+ * Source is provided to this software because we believe users have a     *
+ * right to know exactly what a program is going to do before they run it. *
+ * This also allows you to audit the software for security holes.          *
+ *                                                                         *
+ * Source code also allows you to port DRAKVUF to new platforms, fix bugs, *
+ * and add new features.  You are highly encouraged to submit your changes *
+ * on https://github.com/tklengyel/drakvuf, or by other methods.           *
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <vector>
+
+#include <libinjector/libinjector.h>
+
+#include "private.h"
+#include "procdump.h"
+
+using std::string;
+
+struct vad_info
+{
+    uint32_t type; // TODO Use backed file name instead of type?
+    uint64_t total_number_of_ptes;
+    std::vector<uint64_t> prototype_ptes;
+    uint32_t idx;                       // index in prototype_ptes
+};
+using vad_info_t = struct vad_info;
+using vads_t = std::map<addr_t, vad_info_t>;
+
+struct procdump_ctx
+{
+    vmi_pid_t pid;
+    vmi_pid_t ppid;
+    uint32_t tid;
+    string name;
+    procdump* plugin;
+    drakvuf_trap_t* bp;
+    vads_t vads;
+    x86_registers_t saved_regs;
+    uint64_t idx;
+    addr_t pool;
+    const uint64_t POOL_SIZE_IN_PAGES = 0x100;
+    uint8_t* file;
+    size_t size;
+    size_t offset;
+    string file_path;
+    size_t current_dump_size;
+};
```

Comment:
You should move the structure and function declarations into a `private.h`

Reply:
Done